src/Microsoft.AspNetCore.Routing/Matchers/DfaMatcherBuilder.cs

```diff
@@ -267,54 +286,96 @@ private int AddNode(DfaNode node, List<DfaState> states, List<JumpTableBuilder>
             {
                 // This node has a single transition to but it should accept zero-width segments
                 // this can happen when a node only has catchall parameters.
-                table.DefaultDestination = Transition(node.Parameters);
-                table.ExitDestination = table.DefaultDestination;
+                pathBuilder.DefaultDestination = Transition(node.Parameters);
+                pathBuilder.ExitDestination = pathBuilder.DefaultDestination;
             }
             else if (node.Parameters != null && node.CatchAll != null)
             {
                 // This node has a separate transition for zero-width segments
                 // this can happen when a node has both parameters and catchall parameters.
-                table.DefaultDestination = Transition(node.Parameters);
-                table.ExitDestination = Transition(node.CatchAll);
+                pathBuilder.DefaultDestination = Transition(node.Parameters);
+                pathBuilder.ExitDestination = Transition(node.CatchAll);
             }
             else if (node.Parameters != null)
             {
                 // This node has paramters but no catchall.
-                table.DefaultDestination = Transition(node.Parameters);
+                pathBuilder.DefaultDestination = Transition(node.Parameters);
             }
             else if (node.CatchAll != null)
             {
                 // This node has a catchall but no parameters
-                table.DefaultDestination = Transition(node.CatchAll);
-                table.ExitDestination = table.DefaultDestination;
+                pathBuilder.DefaultDestination = Transition(node.CatchAll);
+                pathBuilder.ExitDestination = pathBuilder.DefaultDestination;
+            }
+
+            if (node.PolicyEdges.Count > 0)
+            {
+                var policyBuilder = new PolicyJumpTableBuilder(node.NodeBuilder);
+                tableBuilders[stateIndex] = (pathBuilder, policyBuilder);
+
+                foreach (var kvp in node.PolicyEdges)
+                {
+                    policyBuilder.AddEntry(kvp.Key, Transition(kvp.Value));
+                }
             }
 
             return stateIndex;
 
             int Transition(DfaNode next)
             {
                 // Break cycles
-                return ReferenceEquals(node, next) ? stateIndex : AddNode(next, states, tables);
+                return ReferenceEquals(node, next) ? stateIndex : AddNode(next, states, tableBuilders);
             }
         }
 
+        // Builds an array of candidates for a node, assigns a 'score' for each
+        // endpoint.
+        internal Candidate[] CreateCandidates(IReadOnlyList<MatcherEndpoint> endpoints)
+        {
+            if (endpoints.Count == 0)
+            {
+                return Array.Empty<Candidate>();
+            }
+            
+            var candiates = new List<Candidate>();
```

Comment:
minor optimization: This creates a candidate for each endpoint right? Pre-initialize the list with the right capacity? Or create an array and assign indexes to avoid ToArray call?

Reply:
sure